### PR TITLE
BUG: Fix join on MultiIndex for mixed Datetimelike and string levels

### LINKED ISSUE #26558: .add gives incorrect result with MI Dataframe with mix of object and datetimes on index.
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import datetime

df1 = pd.DataFrame(
    data = [10],
    index = pd.MultiIndex.from_tuples([('Z', '2019-05-31')]),
    columns= ['A']
)

df2 = pd.DataFrame(
    data = [20],
    index = pd.MultiIndex.from_tuples([('Z', datetime.datetime(2019,5,31))]),
    columns= ['A']
)

df1.add(df2, fill_value=0)

```
#### Problem description

Two multiindexed dataframes, where the inner level is datetime-like string in the first, and actual datetime in the second.

Applying the .add function returns a dataframe with two index lines as expected (one for the string date, and the other for the datetime), but the values in the output do not match the two input frames.

```ipython
df1.add(df2, fill_value=0)
                 A
Z 2019-05-31  20.0
  2019-05-31  20.0
```

#### Expected Output

```ipython
df1.add(df2, fill_value=0)
                 A
Z 2019-05-31  10.0
  2019-05-31  20.0
```

OR

``` ipython
df1.add(df2, fill_value=0)
               A
Z 2019-05-31  30
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.6.final.0
python-bits: 64
OS: Darwin
OS-release: 16.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: None
LOCALE: en_GB.UTF-8

pandas: 0.24.2
pytest: None
pip: 18.1
setuptools: 40.4.3
Cython: None
numpy: 1.16.3
scipy: 1.2.1
pyarrow: None
xarray: None
IPython: 7.5.0
sphinx: None
patsy: None
dateutil: 2.8.0
pytz: 2019.1
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 3.0.3
openpyxl: 2.6.2
xlrd: 1.2.0
xlwt: None
xlsxwriter: 1.1.8
lxml.etree: 4.3.3
bs4: 4.7.1
html5lib: None
sqlalchemy: 1.3.3
pymysql: 0.9.3
psycopg2: None
jinja2: 2.10.1
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by nrebena**
# Problem
From #26558, a minimal example is:
```python
import pandas as pd
import datetime

i1 = pd.Index(['2019-05-31'])
i2 = pd.Index([datetime.datetime(2019, 5, 31)])

mi1 = pd.MultiIndex.from_tuples([('2019-05-31',)])
mi2 = pd.MultiIndex.from_tuples([(datetime.datetime(2019, 5, 31),)])

print(i1.join(i2, return_indexers=True))
print(mi1.join(mi2, return_indexers=True))
   
# (DatetimeIndex(['2019-05-31'], dtype='datetime64[ns]', freq=None), None, None)
# (MultiIndex([('2019-05-31',)], ), None, array([-1]))
```
Here we can see that one of the indexers for multiindex is wrong.
 
# Proposed fix 🩹
The difference of handling for single index is explain by the join method for DatetimeTimedeltaMixin.
I did some refactoring in pandas/core/indexes/datetimelike.py to be able to reuse the same datetime handling method.
Then in the Index.join method, where MultiIndex are handled, I apply this method to each level of the MultiIndex that is a DatetimeTimedeltaMixin.

# PR checkboxes :ballot_box_with_check:  

- [x] closes #26558
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -3408,7 +3409,16 @@ def join(self, other, how="left", level=None, return_indexers=False, sort=False)
 
             # have the same levels/names so a simple join
             if self.names == other.names:
-                pass
+                # For datetimelike levels, we maybe convert the corresponding level
```
Comment (jreback) [inline pandas/core/indexes/base.py]: woa, we don't want to do anything like this.

**Review by jreback — CHANGES_REQUESTED** [review]
your proposed patch is very very specific to DTI/TDI conversions. Please try to narrow down exactly what is going wrong here, this is introducing a massive level of complexity.

**Comment by nrebena** [comment]
> your proposed patch is very very specific to DTI/TDI conversions. Please try to narrow down exactly what is going wrong here, this is introducing a massive level of complexity.

The bug is very specific to DTI/TDI conversions.

What I did here is reproduce what happen when joining a DatetimeLike index with another index, see code here https://github.com/pandas-dev/pandas/blob/23f8763ad6141efe35a30c5abbe3a43f75c6ca8c/pandas/core/indexes/datetimelike.py#L833

The reason this bug happen it that join between DatetimeLike  Index and Object Index works, ie, the returned join_index has casted the object to the corresponding datetimelike, but then ,when calling get_indexer, https://github.com/pandas-dev/pandas/blob/a59fcc9b73868215678724d4107ce970ace92beb/pandas/core/indexes/base.py#L3505, the Object Index return no result for the datetimelike index _(as it should)_.

In brief, in Index.join, for multiindex, we join datetimelike and object together with maybe_casting of the object, but this doesn't happen at the get_indexer level.
When using single Index, the object index is maybe_casted before attempting to join, and the get_indexer then work.

> woa, we don't want to do anything like this.

That is what happen actually for joining DatetimeLikeIndex with ObjectIndex

I am open to suggestion if one could tell me what we expect here.
The actual patch try to make the comportement of multiindex join consistent with the one on single index.
An alternativ is to prevent joining datetimelike with object, and raise by asking to explicitly cast to datetimelike before end, but I suspect this would be a breaking change.

**Comment by jreback** [comment]
@nrebena if indices type are not the same then they must be joined as object

**Comment by nrebena** [comment]
@jreback This is actually not the case for single index, see this example, where the indice type differs, but the resulting type is DatetimeIndex.

```python
import pandas as pd
import datetime

i1 = pd.Index(['2019-05-31'])
i2 = pd.Index([datetime.datetime(2019, 5, 31)])

print(i1.join(i2, return_indexers=True))
# (DatetimeIndex(['2019-05-31'], dtype='datetime64[ns]', freq=None), None, None)
```

Is a solution to fix Index.join to return an object index in this case ?

**Comment by jreback** [comment]
yes see what fixing this breaks (likely would need to depreciate this)

**Comment by nrebena** [comment]
Update on root cause and advice needed.
In the following example we have one datetime index and one string index with a datetime, with their comportment regarding join, intersection, union and equals.
I am not sure what are the desired comportment, but actually, anytime a string can be use as a datetime, it is casted so. See the following result from branch master.
```
import pandas as pd
import datetime

i1 = pd.Index(['2019-05-31'])
i2 = pd.Index([datetime.datetime(2019, 5, 31)])

i2.join(i1, how='outer', return_indexers=True))
# (DatetimeIndex(['2019-05-31'], dtype='datetime64[ns]', freq=None), None, None)
i1.join(i2, how='outer', return_indexers=True) 
# (DatetimeIndex(['2019-05-31'], dtype='datetime64[ns]', freq=None), None, None)

i1.union(i2)
# Index(['2019-05-31', 2019-05-31 00:00:00], dtype='object')
i2.union(i1)                                                                                                 
# Index([2019-05-31 00:00:00, '2019-05-31'], dtype='object')

i1.intersection(i2)
# Index(['2019-05-31'], dtype='object')
i2.intersection(i1)                                                                                          
# DatetimeIndex(['2019-05-31'], dtype='datetime64[ns]', freq=None)

i1.equals(i2)
# True
i2.equals(i1)                                                                                                
# True

date = "2020-03-16"
values = [date, pd.to_datetime(date)]
pd.Categorical(values, ordered=False)
# [2020-03-16, 2020-03-16]
# Categories (1, datetime64[ns]): [2020-03-16]
```

Are this results what should be expected for each one? For the one with unexpected result, what should be the right one?

**Comment by jreback** [comment]
> Update on root cause and advice needed.
> In the following example we have one datetime index and one string index with a datetime, with their comportment regarding join, intersection, union and equals.
> I am not sure what are the desired comportment, but actually, anytime a string can be use as a datetime, it is casted so. See the following result from branch master.
> 
> ```
> import pandas as pd
> import datetime
> 
> i1 = pd.Index(['2019-05-31'])
> i2 = pd.Index([datetime.datetime(2019, 5, 31)])
> 
> i2.join(i1, how='outer', return_indexers=True))
> # (DatetimeIndex(['2019-05-31'], dtype='datetime64[ns]', freq=None), None, None)
> i1.join(i2, how='outer', return_indexers=True) 
> # (DatetimeIndex(['2019-05-31'], dtype='datetime64[ns]', freq=None), None, None)
> 
> i1.union(i2)
> # Index(['2019-05-31', 2019-05-31 00:00:00], dtype='object')
> i2.union(i1)                                                                                                 
> # Index([2019-05-31 00:00:00, '2019-05-31'], dtype='object')
> 
> i1.intersection(i2)
> # Index(['2019-05-31'], dtype='object')
> i2.intersection(i1)                                                                                          
> # DatetimeIndex(['2019-05-31'], dtype='datetime64[ns]', freq=None)
> 
> i1.equals(i2)
> # True
> i2.equals(i1)                                                                                                
> # True
> 
> date = "2020-03-16"
> values = [date, pd.to_datetime(date)]
> pd.Categorical(values, ordered=False)
> # [2020-03-16, 2020-03-16]
> # Categories (1, datetime64[ns]): [2020-03-16]
> ```
> 
> Are this results what should be expected for each one? For the one with unexpected result, what should be the right one?

all of the above should be deprecated. We shouldn't every compare strings and datetimes. We may currently do this, but we need to deprecate it.

**Comment by jreback** [comment]
@nrebena if you would like to update this to the above behavior would be fine; we should not be joining datetimelike things with non-datetimelike things at all.

**Comment by nrebena** [comment]
Sure thing. I'll try to get to it.

**Comment by nrebena** [comment]
I decided to split this, and started another PR here #33531, and will continue this one regarding the behaviour of pd.Categorical.

**Comment by jreback** [comment]
closing this until / unless #33531 is resolved

**Pull request closed without merging**